src/AppInstallerCLICore/CheckpointManager.cpp

```diff
@@ -0,0 +1,221 @@
+
+        // If the context target checkpoint is ahead or equal to the current checkpoint, we have previously executed this state, load checkpoint state from index.
+        // If the context target checkpoint is behind the current checkpoint, we have not yet processed this state, save checkpoint state to index.
+        if (contextCheckpointFlag > checkpointFlag || contextCheckpointFlag == checkpointFlag)
+        {
+            LoadCheckpoint(context, checkpointFlag);
+        }
+        else if (contextCheckpointFlag < checkpointFlag)
+        {
+            SaveCheckpoint(context, checkpointFlag);
+        }
+    }
+
+    bool CheckpointManager::HasContext()
+    {
+        return !m_checkpointIndex->IsEmpty();
+    }
+
+    void CheckpointManager::AddContext(int contextId)
+    {
+        m_checkpointIndex->AddContext(contextId);
+    }
+
+    void CheckpointManager::RemoveContext(int contextId)
+    {
+        m_checkpointIndex->RemoveContext(contextId);
+    }
+
+    std::string CheckpointManager::GetClientVersion()
+    {
+        return m_checkpointIndex->GetClientVersion();
+    }
+
+    std::string CheckpointManager::GetCommandName(int contextId)
+    {
+        return m_checkpointIndex->GetCommandName(contextId);
+    }
+
+    int CheckpointManager::GetFirstContextId()
+    {
+        return m_checkpointIndex->GetFirstContextId();
+    }
+
+    Execution::CheckpointFlag CheckpointManager::GetLastCheckpoint(int contextId)
+    {
+        return static_cast<Execution::CheckpointFlag>(m_checkpointIndex->GetLastCheckpoint(contextId));
+    }
+
+#ifndef AICLI_DISABLE_TEST_HOOKS
+    static bool s_MockCheckpointManagerCleanUp_Override = false;
+
+    void TestHook_MockCheckpointManagerCleanUp_Override(bool status)
+    {
+        s_MockCheckpointManagerCleanUp_Override = status;
+    }
+#endif
+
+    void CheckpointManager::CleanUpIndex()
+    {
+#ifndef AICLI_DISABLE_TEST_HOOKS
+        // Unit tests will handle clean up so this hook is needed to avoid attempting to reset and clean up again.
+        if (s_MockCheckpointManagerCleanUp_Override)
+        {
+            return;
+        }
+#endif
+
+        bool isIndexEmpty = m_checkpointIndex->IsEmpty();
+
+        m_checkpointIndex.reset();
+
+        if (isIndexEmpty)
+        {
+            const auto& checkpointIndexPath = AppInstaller::Repository::Microsoft::CheckpointIndex::GetCheckpointIndexPath(m_checkpointId);
+            if (std::filesystem::remove(checkpointIndexPath))
+            {
+                AICLI_LOG(CLI, Info, << "Checkpoint index deleted: " << checkpointIndexPath);
+            }
+        }
+    }
+
+    CheckpointManager::~CheckpointManager()
+    {
+        CleanUpIndex();
+    }
+
+    void CheckpointManager::SaveCheckpoint(Execution::Context& context, Execution::CheckpointFlag flag)
+    {
+        switch (flag)
+        {
+        case Execution::CheckpointFlag::CommandArguments:
```

Comment:
I think that the arguments should always be stored in the database metadata, so it seems unnecessary to have a flag to specify them.

Also, using a flag to specify what to store isn't particularly scalable here.  Instead, my suggestion is to use a list of context data items that are relevant at the checkpoint.

So the set of data to store is:
1. Automatic metadata: command, command line, COM info, whatever we want or need for debugging or potential resume use.
2. Checkpoint identifier: I would make this a checkpoint name rather than an enum, because a string is fine here and will likely be easier to debug.  The only downside is the potential for collisions, but if we enforce a descriptive name it should help with debugging and reduce the slim chance of that.
3. Execution context data items: Every item that was requested to be stored.  Additionally, I would keep all data from every checkpoint as the workflow progresses for debugging purposes.

The callers would `context.Checkpoint("MyCheckpoint", { Data1, Data2 });` and would be able to inspect the set of checkpoints during resumption and load the one they know about.